### DISCUSSION TITLE
What are different incident strategies provided by ELSA 3.0 when workflow goes in Faulty state

### DISCUSSION THREAD

**Discussion by NiveditaGalwade** [elsa-workflows/elsa-core#5656] [Q&A]
If anything goes wrong or any exception occurs, workflow goes into faulty state. And because of this, we cannot proceed ahead for faulty state. Explain all incident strategies provided by ELSA 3.0 and explain when to use which stragies with example and code snippet according to ELSA 3.0

**Comment by glime-ai** [comment]
In ELSA 3.0, incident strategies determine how the workflow engine should handle errors that occur during the execution of a workflow. There are two primary incident strategies provided out of the box:

1. `FaultStrategy`: This strategy stops the workflow and marks it as faulted when an error occurs. This is useful when you want to halt execution immediately upon encountering an error and potentially trigger some form of alerting or error handling process.

   Example usage:
   ```csharp
   // No specific code snippet provided for FaultStrategy as it is the default behavior.
   ```

2. `ContinueWithIncidentsStrategy`: This strategy allows the workflow to continue executing even when errors occur. It creates an incident record for each error, which can be reviewed later. This is useful when you want the workflow to proceed with its remaining activities despite errors, perhaps logging them for later review or handling non-critical errors that should not stop the entire process.

   Example usage:
   ```csharp
   services.Configure<IncidentOptions>(options =>
   {
       options.DefaultIncidentStrategy = typeof(ContinueWithIncidents);
   });
   ```

   For a specific workflow, you can set the incident strategy like this:
   ```csharp
   public class MyWorkflow : WorkflowBase
   {
       protected override void Build(IWorkflowBuilder builder)
       {
           builder.WorkflowOptions.IncidentStrategyType = typeof(ContinueWithIncidents);
       }
   }
   ```

When to use which strategy:

- Use `FaultStrategy` when you need to ensure that the workflow does not proceed in an inconsistent state or when an error indicates a critical failure that requires immediate attention.

- Use `ContinueWithIncidentsStrategy` when it's acceptable for the workflow to continue running despite errors, such as in scenarios where errors are expected and can be handled later, or when they are not critical to the overall outcome of the workflow.

It's important to note that these strategies can be configured globally for all workflows or specifically for individual workflows. Additionally, custom incident strategies can be implemented by creating a class that implements the `IIncidentStrategy` interface.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).
